example/router/router.hpp

```diff
@@ -0,0 +1,260 @@
+        std::size_t parent_idx{npos};
+
+        // Index of child nodes in the pool
+        detail::child_idx_vector child_idx;
+    };
+
+public:
+    class match_results_base
+    {
+    protected:
+        node const* leaf_ = nullptr;
+        string_view* matches_ = nullptr;
+        string_view* ids_ = nullptr;
+        std::size_t size_ = 0;
+
+        match_results_base() = default;
+
+        explicit
+        match_results_base(
+            node const* leaf,
+            string_view* matches,
+            string_view* ids,
+            std::size_t N)
+            : leaf_(leaf)
+            , matches_(matches)
+            , ids_(ids)
+            , size_(N)
+        {}
+    public:
+        using iterator = string_view*;
+        using const_iterator = string_view const*;
+        using size_type = std::size_t;
+        using difference_type = std::ptrdiff_t;
+        using reference = string_view&;
+        using const_reference = string_view const&;
+        using pointer = string_view*;
+        using const_pointer = string_view const*;
+
+        // result-like functions
+        explicit
+        operator bool() const;
+
+        bool
+        has_value() const;
+
+        bool
+        has_error() const;
+
+        error_code
+        error() const;
+
+        // vector-like functions
+        const_reference
+        at( size_type pos ) const;
+
+        const_reference
+        at( string_view id ) const;
+
+        const_reference
+        operator[]( size_type pos ) const;
+
+        const_reference
+        operator[]( string_view id ) const;
+
+        const_iterator
+        find( string_view id ) const;
+
+        const_iterator
+        begin() const;
+
+        const_iterator
+        end() const;
+
+        bool
+        empty() const noexcept;
+
+        size_type
+        size() const noexcept;
+    };
+
+protected:
+    BOOST_URL_DECL
+    router_base();
+
+    BOOST_URL_DECL
+    void
+    route_impl(
+        string_view s,
+        std::size_t idx,
+        std::size_t maxn);
+
+    BOOST_URL_DECL
+    node const*
+    match_impl(
+        string_view s,
+        string_view*& matches,
+        string_view*& names) const noexcept;
+};
+
+
+/** A URL router.
+
+    This container matches static and dynamic
+    URL requests to an object which represents
+    how the it should be handled. These
+    values are usually callback functions.
+
+    @tparam T type of resource associated with
+    each path template
+
+    @tparam N maximum number of replacement fields
+    in a path template
+
+    @par Exception Safety
+
+    @li Functions marked `noexcept` provide the
+    no-throw guarantee, otherwise:
+
+    @li Functions which throw offer the strong
+    exception safety guarantee.
+
+    @see
+        @ref parse_absolute_uri,
+        @ref parse_relative_ref,
+        @ref parse_uri,
+        @ref parse_uri_reference,
+        @ref resolve.
+*/
+template <class T, std::size_t N = 20>
```

Comment:
LOL no.. the idea was to put the N on the `match_results` not the router class. Although, I'm not sure which is better.

Reply:
Good idea. I thought the proposal was to fix it at 20 and that it was being innovative here. :laughing: 

What would the API for `router<T>::match` look like in this case?

```
template <std::size_t N = 20>
match_results
match(string_view request) const noexcept;
```

```
template <class N = mp_int<20>>
match_results
match(string_view request, N) const noexcept;
```

?

Reply:
ouch... that hurts the eyes...